Update NIP-10 to clarify usage of reposts and quote posts

alexgleason opened the pull request:
Per our discussion in https://github.com/nostr-protocol/nips/issues/173#issuecomment-1449105521, I'm proposing some changes to NIP-10. Note that NIP-10 is still in a draft state.

1. Renamed NIP-10 from "Conventions for clients' use of `e` and `p` tags in text events" to "Replies, Reposts, and Quote Posts in text events". I think this is a more realistic title, and will help implementers find what they're looking for.
2. Removed the "DEPRECATED" section of the document. The document seems like it was originally written to solve a problem that is no longer prevalent.
3. Clarified the meaning of markers, and described how they're used specifically to create reposts and quote posts.
4. Removed the "root" marker. Someone please tell me if there's history I'm missing behind this, but I can't comprehend why a "root" marker would be needed, especially because it's often very difficult or _impossible_ to add it.
5. Clarified that reposts and quote posts MUST reference the event like `#[0]` in the event content.

mikedilger commented:
I think you've gone a bit overboard.

The deprecated section helped me to understand how to interpret events that are created in clients that use the deprecated method. Many events in nostr use that deprecated way of doing things, even new events on old clients or clients that haven't bothered to change, and it is good to have documented how clients should interpret such events, even if clients should not be generating events in that way anymore.

The "root" marker helps clients find the entire thread. If you have that marker, you can get all the events that reference it in a single relay query.  Otherwise you have to go back and forth to the relay getting the parent, then the grandparent, then the great grandparent, etc.   If you find it impossible to add a root marker, then just add a reply marker and clients will have to deal with the lack of information by climbing.  Among a pride of well behaved nostr clients (pride is the collective noun of a group of ostriches) when you reply to an event, that event has the root tag, you just copy it.

I didn't actually read your changes yet, I'm just replying to your comment.

arthurfranca commented:
I've created a [PR](https://github.com/alexgleason/nips/pull/1) to @alexgleason 's branch to fix [#293](https://github.com/nostr-protocol/nips/pull/293) and [#267](https://github.com/nostr-protocol/nips/issues/267). Also added footnote linking to previous versions of NIP-10 because of what @mikedilger pointed out.

fiatjaf commented:
Aside from removing the "root" marker and deleting the deprecated part of the NIP I like the clarifications introduced here.

As @mikedilger said, it is not impossible to add the "root".  If the event you're replying to has no `e` tags it is the root, if it has a "root" tag you copy it, otherwise there is nothing you can do, but so far all clients have been properly adding "root" tags so it's been working.

alexgleason commented:
Thank you both for the feedback, here are the latest changes:

- Added back the "deprecated" section describing positional "e" tags.
- Added back the "root" marker.

Appreciate the explanation about the root marker @mikedilger, makes perfect sense!

monlovesmango commented:
I like these changes, much clearer. 

>Clients MUST NOT include multiple "e" tags with the same marker in an event.

Shouldn't you be able to have multiple 'mention' events tagged?

alexgleason commented:
I updated the document with these changes:

1. Moved "deprecated" section to the bottom, and added additional text about Kind 6.
2. Updated the text to support multiple quotes per post, as pointed out by @monlovesmango

alexgleason commented:
My first instinct on NIP-10 was that you shouldn't be able to mention multiple events, and that it shouldn't be part of the text. When I created this MR I sought to document the system as it currently exists, but after seeing this insanity on my timeline I think we should consider limiting to ONE mentioned event:

![image](https://user-images.githubusercontent.com/3639540/224877570-e042a9ea-4ff5-451d-be0f-5f7f7c1fcf07.png)

Egge7 commented:
Don't mind me just adding my 2 sats here... I really don't like the "subscribe to the root ID to get the thread approach". This will work as long as nostr is early, but with rising adoption this is an incredibly inefficient way to construct a thread. You can even see this today with a "Jack-post" that has hundreds of replies. Suddenly building a thread means fetching multiple megabytes of data... 

In a perfect world a reply would include all its thread event's ids, maybe even with an additional Marker 'ancestor'. A reply would always include it's parents id as a reply e Tag and additionally include all other r tags as ancestors. Then clients could fetch all events, loop through them and build a chain by looking at each events marked 'reply' tag

arthurfranca commented:
@Egge7 you described issue 267. Your opinion (or should I call it display of reality) is important for the repository maintainers to notice that replies are currently flawed.
[Take a look at this](https://github.com/nostr-protocol/nips/issues/267#issuecomment-1455080477) and see how what you say is similar to what I propose. The difference is I came to the conclusion that an extra level/depth tag is a way to keep `e` tags for all ancestors while being able to access any level down the reply tree, not just the direct descendants.

fiatjaf commented:
Then a conversation with 100 back-and-forth responses would get you events with 100 tags, which is horrible for scalability and also a ton of data.

If you're hitting threads with over 100 replies, or even a thousand replies, then you probably need relay-level filtering. No one wants to read 100 replies to any comment, at least not all at once. And 99% of these are worthless "I agree" or "you suck" replies anyway (not to mention the spam). A good client will allow users to filter by relay so they can opt to read only valuable (as deemed by the relay choices) replies or read everything if they really want.

arthurfranca commented:
> Then a conversation with 100 back-and-forth responses would get you events with 100 tags, which is horrible for scalability and also a ton of data

Not true. A leaf (in the reply tree) will only store `e` tags related to the branch it is part of. It won't know about other branches.

For instance, 100 direct replies to the root will start 100 new branches. So such events only have one ancestor, the root one.

If another event replies to one of above 100 replies, it will have only 2 ancestors.

arthurfranca commented:
"What if a branch grows 100 levels deep?" – one may ask

I've already antecipated it in one of my closed PRs. See below
> A relay MAY limit events to a max level. As a recommendation, clients SHOULD keep 4 as a maximum depth (`l` tag as "4") by not appending an `e` tag if already at level "4"

Considering a branch A-B-C-D-E, following above recommendation will simply consider F a reply to D, instead of to E. Could even add a `p` tag with E pubkey making it possible to say "replying to E pubkey" on UI.

monlovesmango commented:
@alexgleason from the looks of your example the problem is nested mentions are pulling in, not that there are multiple mentions

This PR initially was created to just clarify wording and I'd say let's keep to that scope and merge this PR as is to avoid having multiple in flight PR's to the same NIP.

I'm not opposed to changing the methodology of tagging events but I think thats a separate discussion that is being had in PR [We need a way to request direct replies to a note #267](https://github.com/nostr-protocol/nips/issues/267). Lets keep the discussion around tagging events methodology going there so there is a central place for our rationale.

alexgleason commented:
The goal of this PR was not just to improve wording, but to clarify what a repost _is_ and how to create one.

I'm not sure we should merge this until we address this comment: https://github.com/nostr-protocol/nips/issues/173#issuecomment-1478465856

The idea that a repost is defined as a kind 1 note with `"#[0]"` is something I was skeptical of when I first discovered it. After seeing it in action and hearing a little bit of pushback from other people, I think we should reconsider using kind 6 events which are still in common use.

monlovesmango commented:
not opposed to bringing back kind 6, however kind 6 is only for reposts to my understanding. quote posts would still need to be able to use the mention 'e' tag, right?

Egge7 commented:
> not opposed to bringing back kind 6, however kind 6 is only for reposts to my understanding. quote posts would still need to be able to use the mention 'e' tag, right?

Quote-Posts could be kind 6 with a non-empty content field

monlovesmango commented:
not opposed to that. but we would still need `root`, `reply`, and `mention` markers then for kind 6 then.

alexgleason commented:
With NIP-18 being restored, and NIP-27 being added, most of the concerns this MR was trying to address are now resolved.
